File: ecs_posnext/api/sales_by_branch.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals
import json
import frappe
from frappe import _
from frappe.utils import getdate


@frappe.whitelist()
def get_sales_by_branch(filters=None):
    """Return sales breakdown by branch and product."""
    if isinstance(filters, str):
        filters = json.loads(filters or "{}")
    filters = filters or {}

    from_date = filters.get("from_date") or getdate()
    to_date = filters.get("to_date") or getdate()
    branches = _normalize(filters.get("branches"))
    pos_profiles = _normalize(filters.get("pos_profiles"))
    items = _normalize(filters.get("items"))
    item_groups = _normalize(filters.get("item_groups"))
    owners = _normalize(filters.get("owners"))

    branch_expr = _branch_expr()

    base_conditions = ["si.docstatus = 1", "si.posting_date between %(from_date)s and %(to_date)s"]
    params = {"from_date": from_date, "to_date": to_date}

    if branches:
        base_conditions.append(f"{branch_expr} in %(branches)s")
        params["branches"] = tuple(branches)
    if pos_profiles:
        base_conditions.append("si.pos_profile in %(pos_profiles)s")
        params["pos_profiles"] = tuple(pos_profiles)
    item_conditions = []
    if items:
        item_conditions.append("sii.item_code in %(items)s")
        params["items"] = tuple(items)
    if item_groups:
        item_conditions.append("sii.item_group in %(item_groups)s")
        params["item_groups"] = tuple(item_groups)
    if owners:
        base_conditions.append("si.owner in %(owners)s")
        params["owners"] = tuple(owners)

    cond_detail = " AND ".join(base_conditions + item_conditions)

    gross_expr = "COALESCE(sii.base_amount, sii.amount, 0)"
    net_expr = "COALESCE(sii.base_net_amount, sii.net_amount, 0)"
    discount_expr = f"COALESCE({gross_expr} - {net_expr}, 0)"

    detail = frappe.db.sql(
        f"""
        SELECT
            {branch_expr} AS branch,
            si.pos_profile,
            sii.item_code,
            sii.item_name,
            sii.item_group,
            SUM(sii.qty) AS qty,
            SUM({gross_expr}) AS gross_sales,
            SUM({discount_expr}) AS discount_amount,
            SUM({net_expr}) AS net_sales
        FROM `tabSales Invoice` si
        JOIN `tabSales Invoice Item` sii ON sii.parent = si.name
        LEFT JOIN `tabPOS Profile` pp ON pp.name = si.pos_profile
        WHERE {cond_detail}
        GROUP BY {branch_expr}, si.pos_profile, sii.item_code, sii.item_name, sii.item_group
        """,
        params,
        as_dict=True,
    )

    total_net = sum(d.net_sales or 0 for d in detail)

    for d in detail:
        d.tax_amount = d.get("tax_amount") or 0
        d.net_with_tax = (d.net_sales or 0) + d.tax_amount
        d.sales_pct = ((d.net_sales or 0) / total_net * 100) if total_net else 0

    # branch summary
    branch_summary = {}
    for d in detail:
        key = d.branch or "Unknown"
        row = branch_summary.setdefault(
            key,
            {
                "branch": key,
                "gross_sales": 0,
                "discount_amount": 0,
                "tax_amount": 0,
                "net_sales": 0,
                "net_with_tax": 0,
                "qty": 0,
            },
        )
        row["gross_sales"] += d.gross_sales or 0
        row["discount_amount"] += d.discount_amount or 0
        row["tax_amount"] += d.tax_amount or 0
        row["net_sales"] += d.net_sales or 0
        row["net_with_tax"] += d.net_with_tax or 0
        row["qty"] += d.qty or 0

    totals = {
        "gross_sales": sum(r["gross_sales"] for r in branch_summary.values()),
        "discount_amount": sum(r["discount_amount"] for r in branch_summary.values()),
        "tax_amount": sum(r["tax_amount"] for r in branch_summary.values()),
        "net_sales": sum(r["net_sales"] for r in branch_summary.values()),
        "net_with_tax": sum(r["net_with_tax"] for r in branch_summary.values()),
        "qty": sum(r["qty"] for r in branch_summary.values()),
    }

    chart = {
        "labels": [r["branch"] for r in branch_summary.values()],
        "values": [r["net_sales"] for r in branch_summary.values()],
    }

    orders_cond_items = []
    if items:
        orders_cond_items.append("sii2.item_code in %(items)s")
    if item_groups:
        orders_cond_items.append("sii2.item_group in %(item_groups)s")
    cond_orders = " AND ".join(base_conditions + orders_cond_items)

    orders = frappe.db.sql(
        f"""
        SELECT
            si.name,
            si.customer_name,
            {branch_expr} AS branch,
            si.pos_profile,
            si.posting_date,
            COALESCE(si.base_grand_total, si.grand_total, 0) AS grand_total,
            COALESCE(si.base_discount_amount, 0) AS discount_amount,
            COALESCE(si.base_total_taxes_and_charges, 0) AS tax_amount,
            COALESCE(si.base_net_total, si.base_total, 0) AS net_total
        FROM `tabSales Invoice` si
        LEFT JOIN `tabPOS Profile` pp ON pp.name = si.pos_profile
        LEFT JOIN `tabSales Invoice Item` sii2 ON sii2.parent = si.name
        WHERE {cond_orders}
        GROUP BY si.name
        ORDER BY si.posting_date DESC, si.name DESC
        """,
        params,
        as_dict=True,
    )

    return {
        "detail": detail,
        "branch_summary": list(branch_summary.values()),
        "totals": totals,
        "chart": chart,
        "orders": orders,
    }


def _normalize(val):
    if not val:
        return []
    if isinstance(val, str):
        return [v.strip() for v in val.split(",") if v.strip()]
    if isinstance(val, (list, tuple)):
        cleaned = []
        for v in val:
            if isinstance(v, dict):
                cleaned.append(v.get("value") or v.get("name") or v.get("label") or "")
            else:
                cleaned.append(v)
        return [c for c in cleaned if c]
    return [val]


def _branch_expr():
    parts = []
    if frappe.db.has_column("POS Profile", "branch"):
        parts.append("pp.branch")
    if frappe.db.has_column("Sales Invoice", "branch"):
        parts.append("si.branch")
    if not parts:
        return "'Unknown'"
    return f"COALESCE({', '.join(parts)})"

File: ecs_posnext/pos_next/page/sales_by_branch/sales_by_branch.js
frappe.pages["sales_by_branch"].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __("Sales by Branch"),
		single_column: true,
	});

	const $container = $(`
		<div id="sales-by-branch-app" class="py-4"></div>
	`);
	$(page.body).append($container);

	const scripts = [
		"https://cdn.jsdelivr.net/npm/vue@2.7.14/dist/vue.min.js",
		"https://cdn.jsdelivr.net/npm/chart.js",
	];

	_add_styles();
	_load_scripts(scripts)
		.then(() => init_app(page))
		.catch(() => frappe.msgprint(__("Failed to load resources")));
};

function _load_scripts(urls) {
	return Promise.all(
		urls.map(
			(src) =>
				new Promise((resolve, reject) => {
					if (document.querySelector(`script[src="${src}"]`)) return resolve();
					const s = document.createElement("script");
					s.src = src;
					s.onload = resolve;
					s.onerror = reject;
					document.head.appendChild(s);
				})
		)
	);
}

function _add_styles() {
	const css = `
		.sb-filters { display: grid; grid-template-columns: repeat(auto-fit, minmax(160px, 1fr)); gap: 8px; }
		.sb-card { background: #fff; border-radius: 10px; box-shadow: 0 4px 12px rgba(0,0,0,0.05); padding: 12px; }
		.sb-table { width: 100%; border-collapse: collapse; }
		.sb-table th, .sb-table td { padding: 8px 10px; border-bottom: 1px solid #f1f3f5; text-align: left; }
		.sb-table th { background: #f8fafc; text-transform: uppercase; font-size: 11px; letter-spacing: 0.02em; }
		.sb-kpis { display: grid; grid-template-columns: repeat(auto-fit, minmax(180px, 1fr)); gap: 8px; }
	`;
	const tag = document.createElement("style");
	tag.innerHTML = css;
	document.head.appendChild(tag);
}

function init_app(page) {
	/* eslint-disable no-new */
	new Vue({
		el: "#sales-by-branch-app",
		data() {
			const today = frappe.datetime.get_today();
			return {
				loading: false,
				filters: {
					from_date: today,
					to_date: today,
					branches: [],
					pos_profiles: [],
					items: [],
					item_groups: [],
					owners: [],
				},
				totals: {},
				branch_summary: [],
				detail: [],
				orders: [],
				chart: { labels: [], values: [] },
				charts: {},
				ctrls: {},
				showDetails: true,
				showOrders: true,
			};
		},
		computed: {
			kpiCards() {
				return [
					{ key: "net_sales", label: __("Net Sales"), color: "#16a34a" },
					{ key: "net_with_tax", label: __("Net + Tax"), color: "#0ea5e9" },
					{ key: "tax_amount", label: __("Tax"), color: "#f97316" },
					{ key: "discount_amount", label: __("Discount"), color: "#a855f7" },
					{ key: "gross_sales", label: __("Gross Sales"), color: "#64748b" },
					{ key: "qty", label: __("Qty"), color: "#22c55e" },
				].map((c) => ({ ...c, value: this.totals[c.key] || 0 }));
			},
		},
		mounted() {
			this.initControls();
			this.fetchData();
		},
		methods: {
			initControls() {
				this.makeLinkMulti(this.$refs.branchCtrl, "Branch", "branches");
				this.makeLinkMulti(this.$refs.posProfileCtrl, "POS Profile", "pos_profiles");
				this.makeLinkMulti(this.$refs.itemCtrl, "Item", "items");
				this.makeLinkMulti(this.$refs.itemGroupCtrl, "Item Group", "item_groups");
				this.makeLinkMulti(this.$refs.ownerCtrl, "User", "owners", {
					filters: { enabled: 1, user_type: "System User" },
				});
			},
			makeLinkMulti(parentEl, doctype, modelKey, extra = {}) {
				if (!parentEl) return;
				const df = {
					fieldtype: "MultiSelectList",
					label: doctype,
					get_data: function (txt) {
						return frappe.db.get_link_options(doctype, txt, extra.filters || {});
					},
					onchange: () => {},
				};
				const ctrl = frappe.ui.form.make_control({
					df,
					parent: parentEl,
					render_input: true,
				});
				ctrl.$input && ctrl.$input.addClass("form-control");
				const handleChange = () => {
					const raw = ctrl.get_value();
					let vals = [];
					if (Array.isArray(raw)) {
						vals = raw.map((v) => (v && v.value ? v.value : v)).filter(Boolean);
					} else if (typeof raw === "string") {
						vals = raw
							.split(",")
							.map((v) => v.trim())
							.filter(Boolean);
					}
					this.filters[modelKey] = vals;
					this.fetchData();
				};
				// bind to both control change and df onchange
				ctrl.onchange = handleChange;
				df.onchange = handleChange;
				this.ctrls[modelKey] = ctrl;
			},
			formatCurrency(val) {
				return (Number(val) || 0).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 });
			},
			formatNumber(val) {
				return (Number(val) || 0).toLocaleString(undefined, { maximumFractionDigits: 2 });
			},
			makePayload() {
				return {
					from_date: this.filters.from_date,
					to_date: this.filters.to_date,
					branches: this.filters.branches,
					pos_profiles: this.filters.pos_profiles,
					items: this.filters.items,
					item_groups: this.filters.item_groups,
					owners: this.filters.owners,
				};
			},
			fetchData() {
				this.loading = true;
				frappe.call({
					method: "ecs_posnext.api.sales_by_branch.get_sales_by_branch",
					args: { filters: this.makePayload() },
				}).then((r) => {
					const data = r.message || {};
					this.totals = data.totals || {};
					this.branch_summary = data.branch_summary || [];
					this.detail = data.detail || [];
					this.orders = data.orders || [];
					this.chart = data.chart || { labels: [], values: [] };
					this.$nextTick(() => this.drawChart());
				}).finally(() => {
					this.loading = false;
				});
			},
			drawChart() {
				const ctx = document.getElementById("branch-chart");
				if (!ctx || !window.Chart) return;
				if (this.charts.branch) {
					this.charts.branch.destroy();
				}
				this.charts.branch = new Chart(ctx, {
					type: "bar",
					data: {
						labels: this.chart.labels || [],
						datasets: [
							{
								label: __("Net Sales"),
								data: this.chart.values || [],
								backgroundColor: "#0ea5e9",
							},
						],
					},
					options: { responsive: true, scales: { y: { beginAtZero: true } } },
				});
			},
			exportCSV() {
				const rows = [
					["Branch", "POS Profile", "Item Code", "Item", "Item Group", "Qty", "Gross", "Discount", "Tax", "Net", "Net+Tax", "Sales %"],
					...this.detail.map((d) => [
						d.branch || "",
						d.pos_profile || "",
						d.item_code || "",
						d.item_name || "",
						d.item_group || "",
						d.qty || 0,
						d.gross_sales || 0,
						d.discount_amount || 0,
						d.tax_amount || 0,
						d.net_sales || 0,
						d.net_with_tax || 0,
						(d.sales_pct || 0).toFixed(2),
					]),
				];
				const csv = rows.map((r) => r.map((c) => `"${c}"`).join(",")).join("\n");
				const blob = new Blob([csv], { type: "text/csv;charset=utf-8;" });
				const link = document.createElement("a");
				link.href = URL.createObjectURL(blob);
				link.download = "sales_by_branch.csv";
				link.click();
			},
			exportExcel() {
				const rows = [
					["Branch", "POS Profile", "Item Code", "Item", "Item Group", "Qty", "Gross", "Discount", "Tax", "Net", "Net+Tax", "Sales %"],
					...this.detail.map((d) => [
						d.branch || "",
						d.pos_profile || "",
						d.item_code || "",
						d.item_name || "",
						d.item_group || "",
						d.qty || 0,
						d.gross_sales || 0,
						d.discount_amount || 0,
						d.tax_amount || 0,
						d.net_sales || 0,
						d.net_with_tax || 0,
						(d.sales_pct || 0).toFixed(2),
					]),
				];
				const tsv = rows.map((r) => r.join("\t")).join("\n");
				const blob = new Blob([tsv], { type: "application/vnd.ms-excel" });
				const link = document.createElement("a");
				link.href = URL.createObjectURL(blob);
				link.download = "sales_by_branch.xls";
				link.click();
			},
		},
		template: `
			<div class="sb-page">
				<div class="sb-card">
					<div class="sb-filters">
						<input type="date" class="form-control" v-model="filters.from_date" @change="fetchData()" />
						<input type="date" class="form-control" v-model="filters.to_date" @change="fetchData()" />

						<div ref="branchCtrl"></div>
						<div ref="posProfileCtrl"></div>
						<div ref="itemCtrl"></div>
						<div ref="itemGroupCtrl"></div>
						<div ref="ownerCtrl"></div>

						<button type="button" class="btn btn-secondary" @click.prevent="exportCSV()">{{ __('CSV') }}</button>
						<button type="button" class="btn btn-secondary" @click.prevent="exportExcel()">{{ __('Excel') }}</button>
					</div>
				</div>

				<div class="sb-kpis">
					<div v-for="card in kpiCards" :key="card.key" class="sb-card" :style="{borderTop: '3px solid ' + card.color}">
						<div class="text-muted" style="font-size: 11px; text-transform: uppercase;">{{ card.label }}</div>
						<div class="h5 mb-0">{{ ['qty'].includes(card.key) ? formatNumber(card.value) : formatCurrency(card.value) }}</div>
					</div>
				</div>

				<div class="row mt-3">
					<div class="col-md-6 sb-card">
						<h5>{{ __('Branch Summary') }}</h5>
						<table class="sb-table">
							<thead>
								<tr>
									<th>{{ __('Branch') }}</th>
									<th class="text-right">{{ __('Net Sales') }}</th>
									<th class="text-right">{{ __('Tax') }}</th>
									<th class="text-right">{{ __('Net + Tax') }}</th>
								</tr>
							</thead>
							<tbody>
								<tr v-for="row in branch_summary" :key="row.branch">
									<td>{{ row.branch }}</td>
									<td class="text-right">{{ formatCurrency(row.net_sales) }}</td>
									<td class="text-right">{{ formatCurrency(row.tax_amount) }}</td>
									<td class="text-right">{{ formatCurrency(row.net_with_tax) }}</td>
								</tr>
							</tbody>
						</table>
					</div>
					<div class="col-md-6 sb-card">
						<h5>{{ __('Branch Sales Chart') }}</h5>
						<canvas id="branch-chart" height="220"></canvas>
					</div>
				</div>

				<div class="sb-card mt-3">
					<h5 style="cursor:pointer;" @click="showDetails = !showDetails">
						{{ __('Sales Details') }}
						<span class="text-muted" style="font-size:12px;">{{ showDetails ? __('(Hide)') : __('(Show)') }}</span>
					</h5>
					<div v-if="showDetails">
						<table class="sb-table">
							<thead>
								<tr>
									<th>{{ __('Branch') }}</th>
									<th>{{ __('POS Profile') }}</th>
									<th>{{ __('Item Code') }}</th>
									<th>{{ __('Item') }}</th>
									<th>{{ __('Item Group') }}</th>
									<th class="text-right">{{ __('Qty') }}</th>
									<th class="text-right">{{ __('Gross') }}</th>
									<th class="text-right">{{ __('Discount') }}</th>
									<th class="text-right">{{ __('Tax') }}</th>
									<th class="text-right">{{ __('Net') }}</th>
									<th class="text-right">{{ __('Net + Tax') }}</th>
									<th class="text-right">{{ __('Sales %') }}</th>
								</tr>
							</thead>
							<tbody>
								<tr v-for="row in detail" :key="row.branch + row.item_code + row.pos_profile">
									<td>{{ row.branch }}</td>
									<td>{{ row.pos_profile }}</td>
									<td>{{ row.item_code }}</td>
									<td>{{ row.item_name }}</td>
									<td>{{ row.item_group }}</td>
									<td class="text-right">{{ formatNumber(row.qty) }}</td>
									<td class="text-right">{{ formatCurrency(row.gross_sales) }}</td>
									<td class="text-right">{{ formatCurrency(row.discount_amount) }}</td>
									<td class="text-right">{{ formatCurrency(row.tax_amount) }}</td>
									<td class="text-right">{{ formatCurrency(row.net_sales) }}</td>
									<td class="text-right">{{ formatCurrency(row.net_with_tax) }}</td>
									<td class="text-right">{{ (row.sales_pct || 0).toFixed(2) }}%</td>
								</tr>
							</tbody>
						</table>
					</div>
				</div>

				<div class="sb-card mt-3">
					<h5 style="cursor:pointer;" @click="showOrders = !showOrders">
						{{ __('Orders') }}
						<span class="text-muted" style="font-size:12px;">{{ showOrders ? __('(Hide)') : __('(Show)') }}</span>
					</h5>
					<div v-if="showOrders">
						<table class="sb-table">
							<thead>
								<tr>
									<th>{{ __('Order') }}</th>
									<th>{{ __('Customer') }}</th>
									<th>{{ __('Branch') }}</th>
									<th>{{ __('POS Profile') }}</th>
									<th>{{ __('Date') }}</th>
									<th class="text-right">{{ __('Grand Total') }}</th>
									<th class="text-right">{{ __('Discount') }}</th>
									<th class="text-right">{{ __('Tax') }}</th>
									<th class="text-right">{{ __('Net Total') }}</th>
								</tr>
							</thead>
							<tbody>
								<tr v-for="row in orders" :key="row.name">
									<td>{{ row.name }}</td>
									<td>{{ row.customer_name }}</td>
									<td>{{ row.branch }}</td>
									<td>{{ row.pos_profile }}</td>
									<td>{{ row.posting_date }}</td>
									<td class="text-right">{{ formatCurrency(row.grand_total) }}</td>
									<td class="text-right">{{ formatCurrency(row.discount_amount) }}</td>
									<td class="text-right">{{ formatCurrency(row.tax_amount) }}</td>
									<td class="text-right">{{ formatCurrency(row.net_total) }}</td>
								</tr>
							</tbody>
						</table>
					</div>
				</div>
			</div>
		`,
	});
}
